doc/neps/ufunc-overrides.rst

```diff
@@ -106,227 +123,476 @@ Returning ``NotImplemented`` to user should not happen. Moreover::
+incompatible.  Exceptions to this need to check they respect the
+implicit type casting hierarchy.
 
-        without_self = list(inputs)
-        del without_self[pos]
-        without_self = tuple(without_self)
+.. note::
 
-        if func == np.multiply:
-            return self.multiply(*without_self)
+   Note that type casting hierarchy and class hierarchy are here defined
+   to go the "opposite" directions.  It would in principle also be
+   consistent to have ``__array_ufunc__`` handle also instances of
+   subclasses. In this case, the "subclasses first" dispatch rule would
+   ensure a relatively similar outcome. However, the behavior is then less
+   explicitly specified.
+
+Subclasses can be easily constructed if methods consistently use
+:func:`super` to pass through the class hierarchy [7]_.  To support
+this, :class:`ndarray` has its own ``__array_ufunc__`` method,
+equivalent to::
+
+    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+        # Cannot handle items that have __array_ufunc__ (other than our own).
+        outputs = kwargs.get('out', ())
+        for item in inputs + outputs):
+            if (hasattr(item, '__array_ufunc__') and
+                    type(item).__array_ufunc__ is not ndarray.__array_ufunc__):
+                return NotImplemented
 
-        elif func == np.dot:
-            if pos == 0:
-                return self.__mul__(inputs[1])
-            if pos == 1:
-                return self.__rmul__(inputs[0])
-        else:
+        # If we didn't have to support legacy behaviour (__array_prepare__,
+        # __array_wrap__, etc.), we might here convert python floats,
+        # lists, etc, to arrays with
+        # items = [np.asarray(item) for item in inputs]
+        # and then start the right iterator for the given method.
+        # However, we do have to support legacy, so call back into the ufunc.
+        # Its arguments are now guaranteed not to have __array_ufunc__
+        # overrides, and it will do the coercion to array for us.
+        return getattr(ufunc, method)(*items, **kwargs)
+
+Note that, as a special case, the ufunc dispatch mechanism does not call
+this `ndarray.__array_ufunc__` method, even for `ndarray` subclasses
+if they have not overridden the default `ndarray` implementation. As a
+consequence, calling `ndarray.__array_ufunc__` will not result to a
+nested ufunc dispatch cycle.
+
+The use of :func:`super` should be particularly useful for subclasses of
+:class:`ndarray` that only add an attribute like a unit.  In their
+`__array_ufunc__` implementation, such classes can do possible
+adjustment of the arguments relevant to their own class, and pass on to
+the superclass implementation using :func:`super` until the ufunc is
+actually done, and then do possible adjustments of the outputs.
+
+In general, custom implementations of `__array_ufunc__` should avoid
+nested dispatch cycles.  However, for some subclasses, it may be better
```

Comment:
Let's be clear that merely calling `getattr(ufunc, method)(*items, **kwargs)` does not mean "nested dispatch". "Nested dispatch" is when you call `getattr(...)(...)` and catch `TypeError`, as in this example below.